### ISSUE TITLE
Mad Catz's C.T.R.L.R support

### ISSUE THREAD

**Issue by g333no** [ndeadly/MissionControl#113]
The controller can be seen but it's not mapped (shows up with red buttons after pairing) and it also shows up with its proper name in the paired bluetooth devices list. On that note, should i write down the numbers that show up in the btdb.nro for it or send them to you via other means (idk if that's supposed to be private or whatever)?

**Comment by ndeadly** [comment]
Just post a screenshot of the btdb output here. There's nothing in there that can't be shared.

**Comment by g333no** [comment]
![image](https://user-images.githubusercontent.com/50757169/95255269-64062780-0821-11eb-835d-50bc4fb14c4b.png)

**Comment by ndeadly** [comment]
Perfect. Could you run this tool with the controller connected and note down the changes in output for each button press, and the sticks in the extreme X and Y positions? I haven't bothered to optimise performance so what you see on the screen lags by a second or two.

[hid_report_tool.zip](https://github.com/ndeadly/MissionControl/files/5336496/hid_report_tool.zip)

**Comment by g333no** [comment]
sorry if it took a while but the controller has three modes and i mapped all of them
[mad catz ctrlr buttons.txt](https://github.com/ndeadly/MissionControl/files/5336943/mad.catz.ctrlr.buttons.txt)

**Comment by ndeadly** [comment]
Oh wow, you should have mentioned it had multiple modes, you didn't have to do all that haha. You typically want PC/Android mode for these. But thanks anyway. Always good to have more information. 

Give me some time and I'll map the controls and post a build for you to try out.

**Comment by g333no** [comment]
Don't worry, i figured someone else could use the info someday
also i forgot to add but the buttons there follow nintendo's standard, not microsoft's one

**Comment by ndeadly** [comment]
> the buttons there follow nintendo's standard, not microsoft's one

Ah cool thanks for the heads up. A lot of people try to do this, but it just creates confusion. I'd prefer they were just labelled as they are on the controller, and I'll handle the conversion myself. All good though.

**Comment by ndeadly** [comment]
Alright, see how this first attempt goes. I haven't mapped `home` since you were unable to tell me what value it is. I've had this before on some other controllers where it just flashes up for what seems to be a single frame. I think the switch ignores those anyway as I could never get it to work on those controllers. You can use the `MINUS + DPAD_DOWN` combo for it instead.
I didn't bother mapping the media buttons to anything either.

[MissionControl-0.3.0-madcatz.zip](https://github.com/ndeadly/MissionControl/files/5337224/MissionControl-0.3.0-madcatz.zip)

**Comment by g333no** [comment]
> I didn't bother mapping the media buttons to anything either.

I suppose stuff like volume up and down can't be mapped to actually work as volume up and down, right?
Anyway looks like everything is working fine with PC mode, thanks a bunch!

**Comment by ndeadly** [comment]
> I suppose stuff like volume up and down can't be mapped to actually work as volume up and down, right?

As far as I know this info can't be sent from controllers, so at a minimum it's not as straightforward as simply mapping it to a button. There might be some command that could be called to do it, but that's added complexity I'm not ready for just yet. I'll keep it in mind though, could be a useful feature to have.

If you're happy with the controls I'll consider this done. I can try to map home for completeness sake if you're able to determine what value is actually changing, but not guarantee it will work. That said, I made some changes in 0.3.0 that could possibly improve the chances of it working. I don't own any third party controllers that exhibit this behaviour to test with.

**Comment by g333no** [comment]
I tried to look at it and it looks like a 1 (again, first digit of that pair) but i'm not too sure
Anyway i'd say to not bother with it now, everything else works. At best i'll wait for some kind of remapping feature to change the home combo to something else but in the mean time this will do just fine.
Again, thank you so much.

**Comment by ndeadly** [comment]
I would say 1 is likely based on other controllers I've seen, I'll add that for now. Bit confused with your notation there, is it the 3rd or 4th pair along? ie. is it the column with start/select or the dpad values changing?

No problem. Thanks for helping map this controller :)

**Comment by g333no** [comment]
the start/select one, meaning the line should be like this
`0000 | 01 00 10 00 80 7f 80 7f 00 00  |  . . . . . . . . . .  `
if it's actually correct

**Comment by ndeadly** [comment]
Here's one with the changes for home button. It'll either work or it won't. Either way I'll consider this complete for now

[MissionControl-0.3.0-madcatz.zip](https://github.com/ndeadly/MissionControl/files/5337365/MissionControl-0.3.0-madcatz.zip)

**Comment by nono2lozere** [comment]
I've an issue with my controller, it pairs, the buttons works but analog stick doesn't work. Left stick can only go up and right and the right stisk doesn't work at all.
I tried with a PS4 controller and it works flawlessly and my controller works with my PC.

**Comment by ndeadly** [comment]
You might have it connected with the wrong mode. Can you confirm that it's connected with "PC" mode? I believe this is what g333no was using.

I suppose it's also possible that there are different controller firmwares that encode the sticks differently. If it's still not working correctly you can check it with this [improved version](https://github.com/ndeadly/MissionControl/files/5539371/hid_report_viewer.zip) of the tool I posted above. Note down the numeric values shown when you push the sticks to the extremes in X and Y.

**Comment by nono2lozere** [comment]
Thanks for support, Itried both android and PC mode, in android mode there is also issues with the buttons so I think it must be used in PC mode.
I will try your tool this evening when I go back to home and I will report you the results.

**Comment by nono2lozere** [comment]
And I just update the controler firmware from 1.10 to 1.25 I think it's the issue because in changelog there is something about joystick more accurate.
I'll tell you how it work.

**Comment by ndeadly** [comment]
Hopefully that's your issue, though joystick accuracy sounds more like calibration, deadzones etc to me.

**Comment by nono2lozere** [comment]
I tried it and it works perfectly, this was a firmware issue. Menu button doesn't work but the combo does, maybe you can use `PLAY/PAUSE` button instead.
Thanks for your support.

**Comment by ndeadly** [comment]
@nono2lozere glad that solved your issue. Yeah, as noted above, we weren't able to catch what the controller was sending on a press of the home button. I didn't bother taking it any further as all the controllers I've seen that do this (only send a single frame containing the value, no held state) don't seem to work. I think the switch must look for a button being held for X consecutive frames before it considers it pressed. This is one of the reasons I added the combo.

I added your suggestion of `PLAY/PAUSE` as `HOME`. Let me know if it works as expected.
[MissionControl-0.3.0-madcatz.zip](https://github.com/ndeadly/MissionControl/files/5577830/MissionControl-0.3.0-madcatz.zip)
